services/web/server/tests/unit/with_dbs/03/resource_usage/test_usage_services__list.py

```diff
@@ -137,5 +130,147 @@ async def test_list_service_usage(
+
+    # with order by query parameter
+    _filter = {"field": "started_at", "direction": "asc"}
+    url = (
+        client.app.router["list_resource_usage_services"]
+        .url_for()
+        .with_query(order_by=json.dumps(_filter))
+    )
+    resp = await client.get(f"{url}")
+    await assert_status(resp, web.HTTPOk)
+    assert mock_list_usage_services.called
+
+    # with non-supported field in order by query parameter
+    _filter = {"field": "non-supported", "direction": "desc"}
+    url = (
+        client.app.router["list_resource_usage_services"]
+        .url_for()
+        .with_query(order_by=json.dumps(_filter))
+    )
+    resp = await client.get(f"{url}")
+    _, error = await assert_status(resp, web.HTTPUnprocessableEntity)
+    assert mock_list_usage_services.called
+    assert error["status"] == web.HTTPUnprocessableEntity.status_code
+    assert error["errors"][0]["message"].startswith(
+        "We do not support ordering by provided field"
+    )
+
+    # with non-parsable field in order by query parameter
+    url = (
+        client.app.router["list_resource_usage_services"]
+        .url_for()
+        .with_query(order_by=",invalid json")
+    )
+    resp = await client.get(f"{url}")
+    _, error = await assert_status(resp, web.HTTPUnprocessableEntity)
+    assert mock_list_usage_services.called
+    assert error["status"] == web.HTTPUnprocessableEntity.status_code
+    assert error["errors"][0]["message"].startswith("Invalid JSON")
+
+    # with order by without direction
+    _filter = {"field": "started_at"}
+    url = (
+        client.app.router["list_resource_usage_services"]
+        .url_for()
+        .with_query(order_by=json.dumps(_filter))
+    )
+    resp = await client.get(f"{url}")
+    await assert_status(resp, web.HTTPOk)
+    assert mock_list_usage_services.called
+
+    # with wrong direction
+    _filter = {"field": "non-supported", "direction": "wrong"}
+    url = (
+        client.app.router["list_resource_usage_services"]
+        .url_for()
+        .with_query(order_by=json.dumps(_filter))
+    )
+    resp = await client.get(f"{url}")
+    _, error = await assert_status(resp, web.HTTPUnprocessableEntity)
+    assert mock_list_usage_services.called
+    assert error["status"] == web.HTTPUnprocessableEntity.status_code
+    assert error["errors"][0]["message"].startswith(
+        "value is not a valid enumeration member"
+    )
+
+    # without field
+    _filter = {"direction": "asc"}
+    url = (
+        client.app.router["list_resource_usage_services"]
+        .url_for()
+        .with_query(order_by=json.dumps(_filter))
+    )
+    resp = await client.get(f"{url}")
+    _, error = await assert_status(resp, web.HTTPUnprocessableEntity)
+    assert mock_list_usage_services.called
+    assert error["status"] == web.HTTPUnprocessableEntity.status_code
+    assert error["errors"][0]["message"].startswith("field required")
+
+
+@pytest.mark.parametrize("user_role", [(UserRole.USER)])
```

Comment:
Question: do you anywhere test what happens with the different user roles? e.g. anynymous, guest, user, admin ?
basically if there are restrictions and you do not check the different roles anywhere then this is completely untested.

Reply:
Valid point, I have added a test for that, thanks.